dlib/media/ffmpeg_utils.h

```diff
@@ -0,0 +1,403 @@
+                It has a bunch of helper functions for retrieving the frame's properties.
+                See ffmeg documentation for AVFrame.
+                It contains frame data, which may be image, audio or other.
+                The format and layout of the frame is not necessarily RGB (for image) or stereo s16 (for audio).
+                It depends on codec configuration.
+                It is up to the user to appropriately read the data using get_frame().data, get_frame().linesize.
+                There are however helper functions convert() with overloads to convert to/from dlib objects.
+        !*/
+
+        Frame()                         = default;
+        Frame(Frame&& ori)              = default;
+        Frame& operator=(Frame&& ori)   = default;
+        Frame(const Frame& ori);
+        Frame& operator=(const Frame& ori);
+
+        bool            is_empty()      const noexcept;
+        bool            is_image()      const noexcept;
+        bool            is_audio()      const noexcept;
+
+        /*image*/
+        AVPixelFormat   pixfmt()        const noexcept;
+        int             height()        const noexcept;
+        int             width()         const noexcept;
+
+        /*audio*/
+        int             nsamples()      const noexcept;
+        int             nchannels()     const noexcept;
+        uint64_t        layout()        const noexcept;
+        AVSampleFormat  samplefmt()     const noexcept;
+        int             sample_rate()   const noexcept;
+
+        std::chrono::system_clock::time_point get_timestamp() const noexcept;
+        const AVFrame& get_frame() const noexcept;
+
+    private:
+        // Users should only be able to access const functions
+        // However, the implementation details should be able to access non-const functions and data.
+        friend class details::sw_image_resizer;
+        friend class details::sw_audio_resampler;
+        friend class details::sw_audio_fifo;
+        friend class details::decoder_extractor;
+        friend Frame convert(const audio_frame& frame);
+        friend Frame convert(const array2d<rgb_pixel>& frame);
+
+        static Frame make(
+            int h,
+            int w,
+            AVPixelFormat pixfmt,
+            int sample_rate,
+            int nb_samples,
+            uint64_t channel_layout,
+            AVSampleFormat samplefmt,
+            std::chrono::system_clock::time_point timestamp
+        );
+
+        static Frame make_image(
+            int h,
+            int w,
+            AVPixelFormat pixfmt,
+            std::chrono::system_clock::time_point timestamp_us
+        );
+
+        static Frame make_audio(
+            int sample_rate,
+            int nb_samples,
+            uint64_t channel_layout,
+            AVSampleFormat samplefmt,
+            std::chrono::system_clock::time_point timestamp
+        );
```

Comment:
No point in making these private. The encoder/muxer PR (not this one) will likely need to make these public.

Reply:
Also, these can be normal constructors, not named constructors.